### PR TITLE
Wrote get_top_k_trials() fn to return pandas df of top trials of ray tune exp

### LINKED ISSUE #27: Trial reports only show last epoch performance
We want to report performance of the best epoch of the trial, not just the last.

### PR DESCRIPTION
**Pull request by lawrence-chillrud**
Wrote a function to handle the issue that ray tune's experiment reports only show the last epoch performance rather than the best. For the details please see the function documentation!

### CODE REVIEW

**Comment by cooperlab** [comment]
@lawrence-chillrud I think your solution of dealing with experiment files rather than the object returned by `tuner.experiment` is preferable. The files are persistent and can be revisited anytime where the object is transient.

I did some investigation and wanted to mention that it's possible to achieve this using the object. The `scope=all` parameter for [`ray.tune.ResultGrid.get_best_result`](https://docs.ray.io/en/latest/tune/api/doc/ray.tune.ResultGrid.get_best_result.html#ray-tune-resultgrid-get-best-result
) will return the single best epoch of all trials - not just the last epoch result. It is also possible to iterate over the `Result` in the `ResultGrid` object, calling `Result.best_checkpoints` on each to build a longer list. Makes me feel a little better about Ray.

**Comment by cooperlab** [comment]
Should the outputs include the checkpoint directories?

This information is not contained in the results.json files but it can be inferred from the sorting index where you `final_df.sort_values`.

File: glimr/utils.py [hunk 1]
```diff
@@ -1,4 +1,76 @@
 from ray.tune.search import sample
+import os
+import pandas as pd
+
+
+def get_top_k_trials(
+    dir, metric="softmax_balanced", k=10, drop_dups=True, config_filter=None
```
Comment (cooperlab) [inline glimr/utils.py]: Suggest not setting a default value for `metric`.

Alternatively, set default to `None` and infer from .json file as first key.

**Comment by cooperlab** [comment]
Related to this - I don't see a way to report the best epoch in the trial in the reporter. I think it's hardwired to report the current trial state for running trials, and the last epoch for completed trials. For the MNIST example, the last epoch is the best for most trials due to the stopping criteria, but that may not be true for all projects.

I think we can live with this if we have our own analysis tools. The top-k can be run during an experiment so for long-running experiments we can still monitor progress.

File: glimr/utils.py (hunk 1 above)
Comment (lawrence-chillrud) [reply glimr/utils.py]: Incorporated the latter into the function (`metric=None`)

**Comment by lawrence-chillrud** [comment]
> Should the outputs include the checkpoint directories?
> 
> This information is not contained in the results.json files but it can be inferred from the sorting index where you `final_df.sort_values`.

Included the `checkpoint_path`s of the trials in the `final_df` returned by the function -- see column `checkpoint_path`.

Note: if the `metric` specified to `get_top_k_trials` is *not* the same as the metric that was passed to the initial ray tune experiment (i.e., the metric ray tune is optimizing for), then many (or possibly all) of the checkpoints for those trials will not exist, and `None` will be reported rather than a `checkpoint_path`. E.g., if ray tune was told to optimize for balanced accuracy, but then the user passes in AUC as the metric to sort `get_top_k_trials`, trials could be returned in the `final_df` that have no saved checkpoint. However, if in this example the user passes in balanced accuracy as the metric, then a `checkpoint_path` exists for every trial (provided `drop_dups=True`) and will be returned accordingly.

**Review by cooperlab — APPROVED** [review]

**Pull request merged**